File: Gauss.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using Matrix = std::vector<std::vector<double>>;
const std::size_t M = 4;
const std::size_t N = 4;

void printMatr(const Matrix& matr)
{
    for(const auto& i : matr)
    {
        for(const auto& j : i)
        {
            std::cout << j << " ";
        }
        std::cout << '\n';
    }
}

void printVec(const std::vector<double>& vec)
{
    for(const auto& i : vec)
    {
        std::cout << i << " ";

    }
    std::cout << '\n';
}

std::size_t ArgMax(const Matrix& matr, std::size_t i)
{
    double MAX = std::abs(matr[i][i]);
    std::size_t max_index = i;

    for (std::size_t j = i + 1; j < matr.size(); ++j)
    {
        if (std::abs(matr[j][i]) > MAX)
        {
            MAX = std::abs(matr[j][i]);
            max_index = j;
        }
    }
    return max_index;
}

void swapRows(Matrix& matr, std::size_t i, std::size_t j)
{
    for(std::size_t k = 0; k < matr.size(); ++k)
    {
        std::swap(matr[i][k], matr[j][k]);
    }
}


double dot(const std::vector<double>& a, const std::vector<double>& b)
{
    double sum = 0;
    for(std::size_t i = 0; i < a.size(); ++i)
    {
        sum += a[i] * b[i];
    }
    return sum;
}

void GaussMethod(Matrix& A, std::vector<double>& b)
{
    for(std::size_t i = 0; i < A.size() - 1; ++i)
    {
        std::size_t K = ArgMax(A, i);
        swapRows(A, i, K);
        std::swap(b[i], b[K]);
        for(std::size_t j = i + 1; j < A.size(); ++j)
        {
            double factor = A[j][i] / A[i][i];
            for(std::size_t k = 0; k < A.size(); ++k)
            {
                A[j][k] -= factor * A[i][k];
            }
            b[j] -= factor * b[i];
        }
    }
    
    //solution
    std::vector<double> sol(b.size(), 0);
    for(int i = A.size() - 1; i >=0; --i)
    {
        double npdot = dot(A[i], sol);
        
        sol[i] = (b[i] - npdot) / A[i][i];
    }
    std::cout << "Solution: ";
    printVec(sol);
    
}


int main()
{
    Matrix A = {
        {-3,3,1},
        {0,4,-2},
        {-3, 1, 5}
    };
    std::vector<double> B = {-2,5,1};
    GaussMethod(A, B);
}

File: gauss.py
# implementation of the gauss method for solving linear systems Ax = b

import numpy as np
from numpy.linalg import LinAlgError


def gauss(A, b):
    """
    Solve the linear system Ax = b using Gaussian elimination with partial pivoting.

    Parameters:
    A : array_like, shape (n, n)
        Coefficient matrix.
    b : array_like, shape (n,)
        Ordinate or "dependent variable" values.

    Returns:
    x : ndarray, shape (n,)
        Solution to the system Ax = b.

    Raises:
    LinAlgError
        If the matrix A is singular or not square.
    """
    A = np.array(A, dtype=float)
    b = np.array(b, dtype=float)

    n = A.shape[0]
    if A.shape[1] != n or b.size != n:
        raise LinAlgError("Matrix A must be square and vector b must have compatible dimensions.")

    # Forward elimination with partial pivoting
    for k in range(n):
        # Find the pivot row
        max_row_index = np.argmax(np.abs(A[k:n, k])) + k
        if A[max_row_index, k] == 0:
            raise LinAlgError("Matrix is singular.")
        
        # Swap the current row with the pivot row
        if max_row_index != k:
            A[[k, max_row_index]] = A[[max_row_index, k]]
            b[[k, max_row_index]] = b[[max_row_index, k]]

        # Eliminate entries below the pivot
        for i in range(k + 1, n):
            factor = A[i, k] / A[k, k]
            A[i, k:] -= factor * A[k, k:]
            b[i] -= factor * b[k]

    # Back substitution
    x = np.zeros(n)
    for i in range(n - 1, -1, -1):
        if A[i, i] == 0:
            raise LinAlgError("Matrix is singular.")
        x[i] = (b[i] - np.dot(A[i, i + 1:], x[i + 1:])) / A[i, i]

    return x

# Example usage:
A = [[2, 1, -1, 3],
     [1, 3, 2, -1],
     [3, 1, 4, 2],
     [1, -1, 1, 2]]

b = [9, 1, 13, 3]
solution = gauss(A, b)
print("Solution:", solution)
